realm/realm-library/src/main/java/io/realm/Realm.java

```diff
@@ -281,103 +281,141 @@ static Realm createInstance(RealmConfiguration configuration, ColumnIndices[] gl
-            realm.doClose();
-            throw new RealmMigrationNeededException(configuration.getPath(), String.format("Realm on disk need to migrate from v%s to v%s", currentVersion, requiredVersion));
-        }
-        if (currentVersion != UNVERSIONED && requiredVersion < currentVersion && columnIndices == null) {
-            realm.doClose();
-            throw new IllegalArgumentException(String.format("Realm on disk is newer than the one specified: v%s vs. v%s", currentVersion, requiredVersion));
-        }
 
-        // Initialize Realm schema if needed
-        if (columnIndices == null) {
+        if (columnIndices != null) {
+            // copy global cache as a Realm local indices cache
+            realm.schema.columnIndices = columnIndices.clone();
+        } else {
+            final boolean syncingConfig = configuration.isSyncConfiguration();
+
+            if (!syncingConfig && (currentVersion != UNVERSIONED)) {
+                if (currentVersion < requiredVersion) {
+                    realm.doClose();
+                    throw new RealmMigrationNeededException(
+                            configuration.getPath(),
+                            String.format("Realm on disk need to migrate from v%s to v%s", currentVersion, requiredVersion));
+                }
+                if (requiredVersion < currentVersion) {
+                    realm.doClose();
+                    throw new IllegalArgumentException(
+                            String.format("Realm on disk is newer than the one specified: v%s vs. v%s", currentVersion, requiredVersion));
+                }
+            }
+
+            // Initialize Realm schema if needed
             try {
-                initializeRealm(realm);
+                if (!syncingConfig) {
+                    initializeRealm(realm, currentVersion);
+                } else {
+                    initializeSyncedRealm(realm, currentVersion);
+                }
             } catch (RuntimeException e) {
                 realm.doClose();
                 throw e;
             }
-        } else {
-            // copy global cache as a Realm local indices cache
-            realm.schema.columnIndices = columnIndices.clone();
         }
 
         return realm;
     }
 
-    @SuppressWarnings("unchecked")
-    private static void initializeRealm(Realm realm) {
-        long version = realm.getVersion();
-        boolean commitNeeded = false;
-        boolean syncAvailable = realm.configuration.isSyncConfiguration();
+    private static void initializeRealm(Realm realm, long currentVersion) {
+        final boolean unversioned = currentVersion == UNVERSIONED;
 
         try {
-            if (!syncAvailable) {
-                realm.beginTransaction();
-                if (version == UNVERSIONED) {
-                    commitNeeded = true;
-                    realm.setVersion(realm.configuration.getSchemaVersion());
-                }
+            realm.beginTransaction();
```

Comment:
now i remembered why the `commitNeeded` was introduced in the first place. If two thread both start init the Realm, one will win with the transaction lock and get here. the other one will just wait until the transaction is committed. BUT, it still have `unversioned == true`. Which means when it gets the transaction lock, it will happily change everything again in the transaction. So it is VERY IMPORTANT to call `getVersion()` again after the transaction begins! that is the whole point we cannot get rid of the `cancelTransaction` in the functions.